File: src/python/problems/ReverseKGroup.py
# https://leetcode.com/problems/reverse-nodes-in-k-group/
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class ReverseKGroup:
    def Solution(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def Reverse(node, length):
            group = None
            while node and length > 0:
                next = node.next

                node.next = group
                group = node
                node = next
                length -= 1

            return Reverse(group, k - length) if length > 0 else (group, node)

        result = ListNode()
        tail = head
        head = result
        while tail:
            lastElement = tail

            node = Reverse(tail, k)
            tail = node[1]

            head.next = node[0]
            head = lastElement

        return result.next



ReverseKGroup().Solution(ListNode(1, ListNode(2, ListNode(3, ListNode(4)))), 2)

File: src/challenges/csharp/Problems/ReverseKGroup.cs
//https://leetcode.com/problems/reverse-nodes-in-k-group/

namespace LeetCode.Problems;

public sealed class ReverseKGroup : ProblemBase
{
    [Theory]
    [ClassData(typeof(ReverseKGroup))]
    public override void Test(object[] data) => base.Test(data);

    protected override void AddTestCases()
        => Add(it => it.ParamListNode("[1,2,3,4,5]").Param(2).ResultListNode("[2,1,4,3,5]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5]").Param(3).ResultListNode("[3,2,1,4,5]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5,6]").Param(2).ResultListNode("[2,1,4,3,6,5]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5,6]").Param(3).ResultListNode("[3,2,1,6,5,4]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5,6,7,8]").Param(3).ResultListNode("[3,2,1,6,5,4,7,8]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5,6,7,8,9]").Param(3).ResultListNode("[3,2,1,6,5,4,9,8,7]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5,6,7,8]").Param(4).ResultListNode("[4,3,2,1,8,7,6,5]"))
          .Add(it => it.ParamListNode("[1,2,3,4,5]").Param(1).ResultListNode("[1,2,3,4,5]"))
          .Add(it => it.ParamListNode("[1]").Param(1).ResultListNode("[1]"));

    private ListNode Solution(ListNode head, int k)
    {
        (ListNode? group, ListNode? tail) Reverse(ListNode? node, int length)
        {
            ListNode? group = null;
            while (node != null && length > 0)
            {
                var next = node.next;

                node.next = group;
                group = node;
                node = next;
                length--;
            }

            return length > 0 ? Reverse(group, k - length) : (group, node);
        }

        var result = new ListNode();

        var tail = head;
        head = result;
        while (tail != null)
        {
            var lastElement = tail;

            var node = Reverse(tail, k);
            tail = node.tail;

            head.next = node.group;
            head = lastElement;
        }

        return result.next!;
    }
}